### ISSUE TITLE
Remove deprecated twisted.python.zshcomp

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#5767]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)| @thijstriemstra reported|
|-|-|
|Trac ID|trac#5767|
|Type|task|
|Created|2012-07-11 10:27:50Z|
|Branch|https://github.com/twisted/twisted/tree/remove-zshcomp-5767|

`twisted.python.zshcomp` is deprecated as of Twisted 11.1 and can be removed. Shell tab-completion is now handled by `twisted.python.usage`.

<details><summary>Searchable metadata</summary>

```
trac-id__5767 5767
type__task task
reporter__thijs thijs
priority__low low
milestone__ 
branch__branches_remove_zshcomp_5767 branches/remove-zshcomp-5767
branch_author__tomprince tomprince
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1342002470000000 1342002470000000
changetime__1359921383000000 1359921383000000
version__None None
owner__tom_prince tom.prince
cc__thijs cc__teratorn cc__rwall
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)<a name="note_2"></a>|@tomprince commented|
|-|-|

(In [36844]) Branching to remove-zshcomp-5767

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)<a name="note_3"></a>|@tomprince commented|
|-|-|

`twisted/python/zsh` isn't removed, since those files just forward to the new completion mechanism.

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_4"></a>|@wallrj set owner to @wallrj<br>@wallrj set status to `assigned`|
|-|-|

Reviewing...

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_5"></a>|@wallrj set owner to @tomprince<br>@wallrj set status to `new`|
|-|-|

# Code Review

The changes look good. A few minor comments below...

 1. [Build results](https://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/remove-zshcomp-5767)
    1. A few failures but I think they're expected
 2. twisted/topfiles/5767.removal
    1. Typo "deprecated in since"
 3. zsh (zsh 5.0.2 (x86_64-redhat-linux-gnu) completion still seems to work on Fedora 18 with this branch -
    that may be because of the legacy functions mentioned below. I
    found the following...
    ```
    $ find /usr/share/zsh/ -name '*twisted*'
    /usr/share/zsh/5.0.2/functions/_twisted
    ```
 4. twisted/python/zsh
    1. I don't fully understand how zsh completion works, but are you
       sure that these files can't be removed. I notice that they all
       contain the following...
       ```
       THIS DIRECTORY AND ALL FILES INCLUDED ARE DEPRECATED.

       These are the old zsh completion functions for Twisted commands... they used
       to contain full completion functions, but now they've simply been replaced
       by the current "stub" code that delegates completion control to Twisted.

       This directory and included files need to remain for several years in order
       to provide backwards-compatibility with an old version of the Twisted
       stub function that was shipped with Zsh.
       ```
    2.
    3. Is there at least a ticket explaining when these files can be
       removed? I think it would be useful to create one and link to
       it in this README.
    4. teratorn seems to be the maintainer of the zsh completion
       files - maybe he can comment.

+1 from me if you fix the typo in the news file.

If possible open a ticket with a plan and a date for the eventual
removal of the legacy zsh shell functions.

-RichardW.

**Comment by twisted-trac** [comment]
|[<img alt="teratorn's avatar" src="https://avatars.githubusercontent.com/u/154069?s=50" width="50" height="50">](https://github.com/teratorn)<a name="note_6"></a>|@teratorn commented|
|-|-|

Hi,
a few points,

 * Could you indicate in the news file that the *functionality* has not been removed, just the deprecated module?
 * Zsh still ships the old stub in their tree [1]. I've just sent an email (today) [2] asking them to update to the new function. As soon as they do they, we can start a 5 year countdown for removing the twisted/python/zsh/_* files that exist for backwards compatibility with old stub.
 * Folks that PACKAGE twisted, OUGHT to be packaging our Zsh stub function, and installing it as "_twisted" in an appropriate "site-functions" directory for Zsh to find - this obviates the need for zsh to already have a stub function, if the user gets their stub from a Twisted package. I do not know how to communicate any information to packagers of Twisted, or know of any documentation that exists for them (?)

[1] - http://zsh.git.sourceforge.net/git/gitweb.cgi?p=zsh/zsh;a=blob;f=Completion/Unix/Command/_twisted;h=b3b2427290cc56c37d7343cd134fd24b80e41c8c;hb=HEAD
[2] - http://www.zsh.org/mla/workers//2013/msg00079.html

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)<a name="note_7"></a>|@tomprince set status to `closed`|
|-|-|

(In [37058]) Merge remove-zshcomp-5767: Remove deprecated `twisted.python.zshcomp`.

Author: tom.prince
Reviewers: rwall
Fixes: [#5767](https://github.com/twisted/twisted/issues/5767)

Shell tab-completion is now handled by twisted.python.usage.

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)<a name="note_8"></a>|@tomprince commented|
|-|-|

  * The stubs in `twisted/python/zsh` are used with the completion shipped with zsh. They can be removed sometime after distributions are no longer shipping versions of zsh that reference them, if having them becomes a maintenance burden. A ticket could be filed if the later ever comes about.
  * I'll file a bug about creating some documentation for packagers.
